sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsPushDownAggregates.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.expressions.Aggregation;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * A mix-in interface for {@link ScanBuilder}. Data source can implement this interface to
+ * push down aggregates. Depends on the data source implementation, the aggregates may not
+ * be able to push down, partially push down and have final aggregate at Spark, or completely
+ * push down.
+ *
+ * When pushing down operators, Spark pushes down filter to the data source first, then push down
+ * aggregates or apply column pruning. Depends on data source implementation, aggregates may or
+ * may not be able to be pushed down with filters. If pushed filters still need to be evaluated
+ * after scanning, aggregates can't be pushed down.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public interface SupportsPushDownAggregates extends ScanBuilder {
+
+  /**
+   * Pushes down Aggregation to datasource.
+   */
+  AggregatePushDownResult pushAggregation(Aggregation aggregation);
```

Comment:
this is a drive by comment - a colleague showed me this PR, so if it's unhelpful please feel free to ignore. This PR implements pushdown for min/max/sum/count because those map easily to map/reduce primitives on individual values.

I suspect that the interesting thought exercise would be 'how do you push avg?'. In my head, if you can fully push sum and count, you have what you need in order to push avg.

And so, you would likely want to have an optimizer rule that basically says 'if I can push down sum/count into this datasource, convert this avg operation into a sum/count, and then add a 'sum/count' expression that can be computed'.

But here, there's no separation between 'can we push' and 'push' which I _think_ makes this kind of optimization impossible without inefficiency?

Reply:
Sorry, I am not sure if I understand your question correctly. Do you mean how to decide which aggregate can be pushed down and whether the push down has performance gain? I think it's up to the data source implementation to decide.